### PR TITLE
Improve performance of SupplementaryResult + Modify the height of RowDltrUnits when UnitConverter is in Landscape

### LINKED ISSUE #248: "Update rates" button cropped/hidden in some languages when the windows is too small.
**Describe the bug**
The "Update rates" button is cropped/hidden when the textblock "Updated <Datetime>" is wrapped.

image

**Steps To Reproduce**
1. Launch the calculator in french
2. Open the currency converter
3. Resize the window

**Device and Application Information (please complete the following information):**
 - OS Build: 10.0.18348.0
 - Architecture: X64
 - Application Version: 10.1812.10048.0

### PR DESCRIPTION
**Pull request by rudyhuyn**
## Fixes #248 
Before:
image

After:
image

### Description of the changes:
- Use `GridLength::Star` on this control (when all other Grid's children are also `GridLength::Star`) will in fact limit the size available to display the content. With `GridLength::Auto`, the grid will give enough space to display the content (without using too much space when the content is small)

### How changes were validated:
Tested manually with all converters and with different window sizes.

### CODE REVIEW

**Comment by rudyhuyn** [comment]
Disclaimer: it slightly changes the layout, the currency info are displayed closer to the bottom of the window, but it looks pretty good.

It's unfortunately not possible to keep the current layout (2*) and and be sure to display all the info without registering to SizeChanged, etc... If only it was possible to write `<RowDefinition Height="Auto" MinHeight="2*" />`...

Alternative solutions:
- register to SizeChanged and modify the MinHeight when height > 48 🤔
- Increase the MinHeight 👎

**Comment by joseartrivera** [comment]
Great find, thanks for the taking the time to code up a fix! I was able to repro this issue with the English translation also, although it was a bit harder to get to. 

I did notice that this also changes the positioning of the "About equal to" text in other converter modes. The lack of padding with the bottom of the window looks a bit odd. @grochocki do you know of any guidelines around padding that we should take into consideration? Added screen shot below, right side is with the fixes from this PR.

image

**Comment by grochocki** [comment]
@joseartrivera The left screenshot is the desired layout/spacing, so any fix to clipping should not change that layout when there would otherwise be no clipping.

For reference, you typically want to work in padding increments of 4px. We have pretty tight gutter padding app-wide, so we should be consistent, but 12px or 24px would be typical for window edges like this.

**Comment by rudyhuyn** [comment]
I usually don't like using `SizeChanged` when we can keep everything in the XAML, but `ActualHeight` not being a dependency property and `RowDefinition::MinHeight` not being a GridLength, it's the only solution we have.

**Comment by joseartrivera** [comment]
Thanks for fixing the formatting Rudy. I am seeing some clipping now with the latest changes on the "About equal to" values now. I will try to brainstorm this weekend to see if we can figure out a good solution for all of these clipping issues.

image

**Comment by rudyhuyn** [comment]
> Thanks for fixing the formatting Rudy. I am seeing some clipping now with the latest changes on the "About equal to" values now. I will try to brainstorm this weekend to see if we can figure out a good solution for all of these clipping issues.
> 
> image

Interesting, probably because of the HorizontalAlignement="Left" (to decrease a little the number of `SizeChanged` event sent when users resize the window), I can push an improved version late tonight

**Comment by rudyhuyn** [comment]
I think SupplementaryResult really need an update, the way the control
manages items, especially OnSupplementaryValuesLayoutUpdated is far from
ideal. I will update this PR with a brand new control asap.

**Comment by rudyhuyn** [comment]
Refactoring of `SupplementaryResult` done. The current implementation had some design flaws:
- It relied on the parent of the UserControl to compute the number of items 
```
Grid^ parentGrid = dynamic_cast<Grid^>(VisualTreeHelper::GetParent(this));	
    if (parentGrid != nullptr)	
    {	
        double parentWidth = parentGrid->ActualWidth; // Bad
```
- It completely refreshed items (and redraw) when the window was resized (bad performance)
```
void SupplementaryResults::OnWindowSizeChanged(Platform::Object^ sender, Windows::UI::Core::WindowSizeChangedEventArgs^ e)	    if (suppResult == nullptr)
{	        return false;
    // to reload supplementary results every time the window is resized
    RefreshData();	
}	
```
- It didn't work if all the parent controls were not `HorizontalAlignment = "Stretch"`
- It created a copy of list and then removed items from the itemsource instead of managing the visibility UI-side.

The new control relies on a new Panel managing alone the visibility of items depend of the size available and not on the size of the parent. This new panel is generic, so it can also be used in other part of the app.

Bonus: the control will try to display the maximum of information, replacing if necessary the last (or the one before) item by a shorter one if the space available isn't enough. See GIFs:

**Before: (the control will only remove the last items)**
D7524D3E-D850-42DB-AE08-103348EBC9A7

**After: (the control will replace the last items by shorter ones and then remove them**
CAF5B486-9D82-4A38-B40D-951A24C949EE

**Comment by danbelcher-MSFT** [comment]
Hey @rudyhuyn, not sure if this is something you can fix with your change, but there are still some window sizes in landscape orientation where the hyperlink is cut:
image
500h x 476w

Additionally, (and, again, I'm not expecting you to fix with this change), there are certain textblocks that only show in Currency Converter for failure cases, such as failing to fetch new currency rates and when the user is offline:
image
image

It's important to consider these edge cases when changing the UI for Currency Converter.

**Comment by rudyhuyn** [comment]
@danbelcher-MSFT 

I don't have this issue when I test, can you send me instructions to repro it?
image


The second case is already managed and tested:
image

The first case can still be buggy, but it will be an extra PR to replace the `HyperlinkButton` CurrencyRefreshBlock by a `Hyperlink` in the Textblock used by `CurrencySecondaryStatus` so it will wrap correctly.

File: src/Calculator/Calculator.vcxproj [hunk 1]
```diff
@@ -390,6 +391,7 @@
     <ClCompile Include="Converters\ItemSizeToVisibilityConverter.cpp" />
     <ClCompile Include="Converters\RadixToStringConverter.cpp" />
     <ClCompile Include="Converters\VisibilityNegationConverter.cpp" />
+    <ClCompile Include="Controls\HorizontalNoOverflowStackPanel.cpp" />
```
[the remark this answers is not in the document]
Comment (rudyhuyn) [reply src/Calculator/Calculator.vcxproj]: I'm a little confused, what's the purpose of this filter when it already matches the real folder structure?

[the remark this answers is not in the document]
Comment (joseartrivera) [reply src/Calculator/Calculator.vcxproj]: It will help VisualStudio organize the files into the Controls folder when viewing from the solutions view instead of the files view. To be honest, I am not sure why this isn't done automatically, but you can make this change by just dragging the new files into the controls folder from VS.

image

**Comment by danbelcher-MSFT** [comment]
> @danbelcher-MSFT
> I don't have this issue when I test, can you send me instructions to repro it?

Hmm, I can't repro now.  I think I was on the wrong commit.  My apologies; I'm still getting used to the transition to GitHub.  Feel free to ignore my previous post!

File: src/Calculator/Controls/HorizontalNoOverflowStackPanel.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        if (item != lastChild)
+        {
+            widthAvailable -= lastChildWidth;
+        }
+        float itemWidth = item->DesiredSize.Width;
+        if (widthAvailable > 0 && itemWidth <= widthAvailable)
+        {
+            //stack the items horizontally (left to right)
+            item->Arrange(Rect(posX, 0, itemWidth, finalSize.Height));
+            posX += item->DesiredSize.Width;
```
[the remark this answers is not in the document]
Comment (rudyhuyn) [reply src/Calculator/Controls/HorizontalNoOverflowStackPanel.cpp]: I was not sure if `Arrange` would change the value of `Desired.Width`. But I probably should use `RenderSize.Width` instead.

File: src/Calculator/Views/UnitConverter.xaml.cpp [hunk 3]
```diff
@@ -385,3 +385,10 @@ void UnitConverter::HideProgressRing()
 
     CurrencyLoadingProgressRing->IsActive = false;
 }
+
+
+void CalculatorApp::UnitConverter::SupplementaryResultsPanelInGrid_SizeChanged(Platform::Object^ sender, Windows::UI::Xaml::SizeChangedEventArgs^ e)
+{
+    //We add 0.01 to be sure to not create an infinite loop with SizeChanged events cascading due to float approximation 
+    RowDltrUnits->MinHeight = max(48, e->NewSize.Height + 0.01);
```
[the remark this answers is not in the document]
Comment (rudyhuyn) [reply src/Calculator/Views/UnitConverter.xaml.cpp]: MinHeight is a dependency property, it already checks for value changes

File: src/Calculator/Calculator.vcxproj (hunk 1 above)
[the remark this answers is not in the document]
Comment (rudyhuyn) [reply src/Calculator/Calculator.vcxproj]: TIL the Solution Explorer use Filters but not folder structure when `Show All Files` is off (but use the Folder structure when it's on), weird behavior! I will change that!

File: src/Calculator/Views/UnitConverter.xaml.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (joseartrivera) [reply src/Calculator/Views/UnitConverter.xaml.cpp]: Good call

File: src/Calculator/Controls/HorizontalNoOverflowStackPanel.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+bool HorizontalNoOverflowStackPanel::ShouldPrioritizeLastItem()
+{
+    return false;
+}
+
+Size HorizontalNoOverflowStackPanel::ArrangeOverride(Size finalSize)
+{
+    float posX = 0;
+    if (Children->Size == 0)
```
[the remark this answers is not in the document]
Comment (joseartrivera) [reply src/Calculator/Controls/HorizontalNoOverflowStackPanel.cpp]: Can you update this on line 44 in this file and in the ShouldPrioritizeLastItem function in SupplementaryResults.xaml.cpp?

[the remark this answers is not in the document]
Comment (rudyhuyn) [reply src/Calculator/Controls/HorizontalNoOverflowStackPanel.cpp]: Done!

**Comment by joseartrivera** [comment]
Closing PR and reopening to kick off builds

**Comment by HowardWolosky** [comment]
@rudyhuyn  - can you resovle the merge conflicts in this PR?  Then I believe we'll be able to merge it in.  Thanks!

**Comment by rudyhuyn** [comment]
> @rudyhuyn - can you resovle the merge conflicts in this PR? Then I believe we'll be able to merge it in. Thanks!

I will tonight!
